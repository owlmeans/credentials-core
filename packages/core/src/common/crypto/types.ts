/**
 *  Copyright 2023 OwlMeans, Inc
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import type { encodeBase58, decodeBase58, toBeArray, getBytes } from 'ethers/lib.esm/utils'
import type { sha256, randomBytes } from 'ethers/lib.esm/crypto'
import type { HDNodeWallet } from 'ethers/lib.esm/wallet'
import type { secp256k1 } from '@noble/curves/secp256k1'

export type CryptoHelper = {
  buildSignSuite: (keyOptions: BuildSignSignatureOptions) => Object

  hash: (data: Buffer | string) => string

  hashBytes: (data: Buffer) => string

  sign: (data: string | Uint8Array, key: string) => string

  getRandomBytes: (size: number) => Uint8Array

  verify: (signature: string, data: string | Uint8Array, key: string) => boolean

  normalizePassword: (password: string) => Uint8Array

  encrypt: (body: string, password: string) => Promise<string>

  decrypt: (chiper: string, password: string) => Promise<string>

  makeDerivationPath: (index?: number, change?: number, account?: number, bc?: string) => string

  base58: () => Base58Lib

  makeId: (key: string, payload?: string, expand?: boolean) => string

  /**
   * @bug seed param implicitly requires Buffer not Uint8Array!
   */
  getKey: (seed: Uint8Array, derivationPath?: string) => CryptoKey & { dp: string }
}

export interface CryptoAdapter {
  base58: {
    encode: typeof encodeBase58
    decode: typeof decodeBase58
    toArray: typeof toBeArray
  }

  sha256: {
    toBytes: typeof getBytes
    hash: typeof sha256
  }

  aes: {
    encoder: any
  }

  secp: {
    sign: typeof secp256k1.sign,
    verify: typeof secp256k1.verify
  },

  random: typeof randomBytes

  WalletClass: typeof HDNodeWallet

  _wallets: {
    [key: string]: HDNodeWallet
  }

  setBase58Impl: (encode: typeof encodeBase58, decode: typeof decodeBase58, toArray: typeof toBeArray) => void
  setSha256Impl: (hash: typeof sha256, toBytes: typeof getBytes) => void
  setAesImpl: (encoder: any) => void
  setRandomImpl: (random: typeof randomBytes) => void
  setSecpImpl: (sign: typeof secp256k1.sign, _verify: typeof secp256k1.verify) => void
}

export type Base58Lib = {
  encode: (a: Uint8Array) => string
  decode: (a: string) => Uint8Array
}

export type BuildSignSignatureOptions = {
  publicKey: string
  privateKey: string
  id: string
  controller: string
}

export type CryptoKey = {
  id?: string
  pk?: string
  pubKey?: string
  nextKeyDigest?: string,
  fragment?: string
}

export const COMMON_CRYPTO_ERROR_NOPK = 'COMMON_CRYPTO_ERROR_NOPK'
export const COMMON_CRYPTO_ERROR_NOPUBKEY = 'COMMON_CRYPTO_ERROR_NOPUBKEY'
export const COMMON_CRYPTO_ERROR_NOID = 'COMMON_CRYPTO_ERROR_NOID'
export const COMMON_CRYPTO_ERROR_ISNOTFULL = 'COMMON_CRYPTO_ERROR_ISNOTFULL'
export const COMMON_CRYPTO_ERROR_NOKEY = 'COMMON_CRYPTO_ERROR_NOKEY'