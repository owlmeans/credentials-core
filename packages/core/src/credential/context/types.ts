import { TContext } from "@affinidi/vc-common"
import { Validatied } from "@affinidi/vc-common/dist/verifier/util"

import { CryptoHelper, CommonCryptoKey } from "@owlmeans/regov-ssi-common"
import { KeyChainWrapper } from "../../keys/types"
import {
  CommonCredentail,
  CommonCredentailSubject,
  CommonSubjectType,
  CommonType,
  CommonUnsignedCredential
} from "./types/credential"
import { DIDDocument, DIDRegistryWrapper } from '@owlmeans/regov-ssi-did'
import { CommonPresentation, CommonPresentationHolder, CommonUnsignedPresentation } from "./types/presentation"


export type BuildCommonContextMethod = (options: {
  keys: KeyChainWrapper
  crypto: CryptoHelper
  did: DIDRegistryWrapper
}) => Promise<CommonContext>

export type CommonContext = {
  keys: KeyChainWrapper
  crypto: CryptoHelper
  did: DIDRegistryWrapper
  buildCredential: CommonBuildCredentialMethod
  signCredential: CommonSignCredentialMethod
  verifyCredential: CommonVerfiyCredentailMethod
  buildPresentation: CommonBuildPresentationMethod
  signPresentation: CommonSignPresentationMethod
}

export type CommonBuildCredentialMethod = <
  SubjectType extends CommonSubjectType = CommonSubjectType,
  Subject extends CommonCredentailSubject<SubjectType> = CommonCredentailSubject<SubjectType>
  >(options: CommonBuildCredentailOptions<SubjectType>) =>
  Promise<CommonUnsignedCredential<Subject>>

export type CommonContextType = TContext

export type CommonBuildCredentailOptions<
  SubjectType extends CommonSubjectType = CommonSubjectType,
  Subject extends CommonCredentailSubject<SubjectType> = CommonCredentailSubject<SubjectType>
  > = {
    id: string,
    type: CommonType
    holder: string,
    subject: Subject,
    issueanceDate?: string
    context: CommonContextType
  }

export type CommonSignCredentialMethod = <
  Subject extends CommonCredentailSubject = CommonCredentailSubject
  >(
  unsingedCredential: CommonUnsignedCredential<Subject>,
  issuer: string,
  key: CommonCryptoKey,
  options?: CommonSignCredentialOptions
) =>
  Promise<CommonCredentail<Subject>>

export type CommonSignCredentialOptions = {
  controllerRole?: ControllerRole
  buildProofPurposeOptions?: () => Promise<Object>
}

export type CommonVerfiyCredentailMethod = (
  credential: CommonCredentail,
  key: CommonCryptoKey
) => Promise<[boolean, CommonVerificationResult]>

export type CommonVerificationResult<Credential extends CommonCredentail = CommonCredentail> = Validatied<Credential>

export type CommonBuildPresentationMethod = <
  Credential extends CommonCredentail = CommonCredentail,
  Holder extends CommonPresentationHolder = CommonPresentationHolder
  >(
  credentials: Credential[],
  options: CommonBuildPresentationOptions<Holder>
) => Promise<CommonUnsignedPresentation<Credential, Holder>>

export type CommonBuildPresentationOptions<
  Holder extends CommonPresentationHolder = CommonPresentationHolder
  > = {
    type?: string | string[]
    context?: CommonContextType
    holder: Holder | DIDDocument
  }

export type CommonSignPresentationMethod = <
  Credential extends CommonCredentail = CommonCredentail,
  Holder extends CommonPresentationHolder = CommonPresentationHolder
  >(
  unsignedPresentation: CommonUnsignedPresentation<Credential, Holder>,
  holder: DIDDocument,
  key: CommonCryptoKey,
  options?: CommonSignPresentationOptions
) => Promise<CommonPresentation<Credential, Holder>>

export type CommonSignPresentationOptions = {
  buildProofPurposeOptions?: () => Promise<Object>
  challange?: string
  domain?: string
}

export type CommonVerifyPresentationMethod = (
  presentation: CommonPresentation,
  key: CommonCryptoKey
) => Promise<[boolean, CommonVerifyPresentationResult]>

export type CommonVerifyPresentationResult<
  Presentation extends CommonPresentation = CommonPresentation
  > = Validatied<Presentation>

export type ControllerRole =
  typeof COMMON_CONTROLLER_ROLE_ISSUER
  | typeof COMMON_CONTROLLER_ROLE_HOLDER

export const COMMON_CONTROLLER_ROLE_ISSUER = 'issuer'
export const COMMON_CONTROLLER_ROLE_HOLDER = 'holder'