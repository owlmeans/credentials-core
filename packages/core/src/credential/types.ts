import { CommonContextType } from "./context/types"
import {
  CommonCredentail,
  CommonCredentailSubject,
  CommonSubjectType,
  CommonType,
  CommonUnsignedCredential
} from "./context/types/credential"
import { CommonPresentation, CommonPresentationHolder, CommonPresentationType, CommonUnsignedPresentation } from "./context/types/presentation"


export type Credential<Subject extends CredentialSubject = CredentialSubject>
  = CommonCredentail<Subject> & {}

export type Identity<Subject extends IdentitySubject = IdentitySubject>
  = {} & Credential<Subject>

export type IdentitySubject<Type extends CredentialSubjectType = CredentialSubjectType> = CredentialSubject<Type>

export type CredentialSubjectProperty<Type extends CredentialSubject = CredentialSubject>
  = Type

export type CredentialSubject<SubjectType extends CredentialSubjectType = CredentialSubjectType> =
  CommonCredentailSubject<SubjectType> & {}

export type CredentialSubjectType = CommonSubjectType & {}

export type CredentialContextType = CommonContextType & {}

export type UnsignedCredentail<
  Subject extends CredentialSubject = CredentialSubject
  > = CommonUnsignedCredential<Subject> & {}

export type CredentialType = CommonType

export type UnsignedPresentation<
  C extends Credential = Credential,
  H extends PresentationHolder = PresentationHolder,
  > = CommonUnsignedPresentation<C, H>

export type Presentation<
  C extends Credential = Credential,
  H extends PresentationHolder = PresentationHolder
  > = CommonPresentation<C, H>

export type PresentationHolder = CommonPresentationHolder

export type PresentationType = CommonPresentationType