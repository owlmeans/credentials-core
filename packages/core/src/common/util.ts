/**
 *  Copyright 2022 OwlMeans
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { MaybeArray, Idish } from "./types"

export const normalizeValue = <Type>(value: MaybeArray<Type>): Type[] => {
  return Array.isArray(value)
    ? value
    : value ? [value] : []
}

export const simplifyValue = <Type>(value: MaybeArray<Type>) => {
  return Array.isArray(value) && value.length < 2
    ? value.length === 0 ? undefined : value[0]
    : value
}

export const singleValue = <Type>(value: MaybeArray<Type>): Type | undefined => {
  return Array.isArray(value)
    ? value.length > 0 ? value[0] : undefined
    : value
}

export const mapValue = async <Type, Res>(
  value: MaybeArray<Type>,
  callback: (value: Type) => Promise<Res>
): Promise<Res[]> => {
  return Promise.all(normalizeValue(value).map(callback))
}

export const addToValue = <Type>(value: MaybeArray<Type> | undefined, add: MaybeArray<Type>): MaybeArray<Type> => {
  if (value) {
    return [...normalizeValue(value), ...normalizeValue(add)]
  }

  return add
}

export const extractId = (id: Idish) => typeof id === 'string' ? id : id.id

export const getDeepValue = <Type, Subject extends {}>(subject: Subject, key: string, def?: Type): Type => {
  const parts = key.split('.')
  return parts.reduce((value, key) => {
    if (subject.hasOwnProperty(key)) {
      const tmp = (subject as any)[key]
      subject = tmp

      return tmp
    }

    return def
  }, def) as Type
}