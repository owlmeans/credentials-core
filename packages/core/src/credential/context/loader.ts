import { DIDRegistryWrapper } from "@owlmeans/regov-ssi-did";


export const getDocumentLoader = (did: DIDRegistryWrapper) =>
  async (url: string): Promise<any> => {
    if (url.startsWith('did:')) {
      return {
        contextUrl: null,
        document: (did => {
          const newDid = JSON.parse(JSON.stringify(did))
          /**
           * @TODO It looks like theese guys doesn't conisdere more
           * sophistacted did structure... or I understand it wrong :)
           * Nevertheless, it should be fixed somewhere else
           */
          if (newDid.proof?.controller) {
            newDid.id = newDid.proof?.controller
          }

          return newDid
        })(await did.lookUpDid(url)),
        /**
          document: {
            '@context': [
              'https://w3id.org/security/v2',
              'https://w3id.org/did/v1'
            ],
            id: url,
            publicKey: [{
              id: `${url}#primary`,
              usage: 'signing',
              type: 'Secp256k1VerificationKey2018',
              publicKeyHex: Buffer.from(crypto.base58().decode(url.split(':')[2])).toString('hex')
            }],
            assertionMethod: [`${url}#primary`],
            authentication: [`${url}#primary`],
          },
         */
        documentUrl: url,
      }
    }

    const jsonld = require('jsonld')

    return jsonld.documentLoader(url)
  }