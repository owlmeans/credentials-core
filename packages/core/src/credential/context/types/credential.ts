import { MaybeArray, SimpleThing, VCV1, VCV1Subject, VCV1Type, VCV1Unsigned } from "@affinidi/vc-common";

export type CommonSubjectType<ExtendedData extends {} = {}> = SimpleThing & ExtendedData

export type CommonCredentailSubject<
  SubjectType extends CommonSubjectType = CommonSubjectType,
  ExtendedType extends {} = {}
  >
  = VCV1Subject<SubjectType> & ExtendedType

export type CommonUnsignedCredential<
  Subject extends MaybeArray<CommonCredentailSubject> = MaybeArray<CommonCredentailSubject>
  > = VCV1Unsigned<Subject>

export type CommonCredential<
  Subject extends MaybeArray<CommonCredentailSubject> = MaybeArray<CommonCredentailSubject>
  > = VCV1<Subject>

export type CommonType = VCV1Type