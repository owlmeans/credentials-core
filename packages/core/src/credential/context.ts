
import { buildVCV1, buildVCV1Skeleton, buildVCV1Unsigned, buildVPV1Unsigned, validateVCV1 } from "@affinidi/vc-common"

import { BuildCommonContextMethod, CommonBuildCredentailOptions, CommonSignCredentialOptions } from "./context/types"
import { CommonCredentail, CommonCredentailSubject, CommonSubjectType, CommonUnsignedCredential } from "./context/types/credential"
import {
  COMMON_CRYPTO_ERROR_NOID,
  COMMON_CRYPTO_ERROR_NOPK,
  COMMON_CRYPTO_ERROR_NOPUBKEY,
  CommonCryptoKey,
  basicHelper
} from "@owlmeans/regov-ssi-common"
import { CommonPresentationHolder, CommonUnsignedPresentation } from "./context/types/presentation"
import { CommonBuildPresentationOptions } from '.'
import { getDocumentLoader } from "./context/loader"

/**
 * @TODO Sign and verify VC with nonce from did.
 * Probably it can be done on the subject level
 */
export const buildCommonContext: BuildCommonContextMethod = async ({
  keys,
  crypto,
  did
}) => {
  const documentLoader = getDocumentLoader(did)

  return {
    keys,

    crypto,

    did,

    buildCredential: async <
      T extends CommonSubjectType = CommonSubjectType,
      S extends CommonCredentailSubject<T> = CommonCredentailSubject<T>
    >(options: CommonBuildCredentailOptions<T>) => {
      const skeleton = buildVCV1Skeleton({
        context: options.context,
        id: options.id,
        type: options.type,
        holder: {
          id: options.holder
        },
        credentialSubject: options.subject,
      })

      return buildVCV1Unsigned({
        skeleton,
        issuanceDate: options.issueanceDate || (new Date).toISOString()
      }) as CommonUnsignedCredential<S>
    },

    signCredential: async <
      S extends CommonCredentailSubject = CommonCredentailSubject
    >(
      unsingedCredential: CommonUnsignedCredential<S>,
      issuer: string,
      key: CommonCryptoKey,
      options?: CommonSignCredentialOptions
    ) => {
      if (!key.id) {
        throw new Error(COMMON_CRYPTO_ERROR_NOID)
      }
      if (!key.pk) {
        throw new Error(COMMON_CRYPTO_ERROR_NOPK)
      }

      try {
        return await buildVCV1({
          unsigned: unsingedCredential,
          issuer: {
            did: issuer,
            keyId: key.fragment || 'publicKey-1',// @TODO Get it from some input
            privateKey: key.pk,
            publicKey: key.pubKey
          },
          getSignSuite: (options) => {
            return crypto.buildSignSuite({
              publicKey: <string>options.publicKey,
              privateKey: options.privateKey,
              id: `${options.controller}#${options.keyId}`,
              controller: options.controller
            })
          },
          documentLoader,
          getProofPurposeOptions: options?.buildProofPurposeOptions
        }) as CommonCredentail<S>
      } catch (e: any) {
        console.log(e.details)

        throw e
      }
    },

    verifyCredential: async (credential, key: CommonCryptoKey) => {
      const result = await validateVCV1({
        getVerifySuite: (options) => {
          if (!key.id) {
            throw new Error(COMMON_CRYPTO_ERROR_NOID)
          }
          if (!key.pubKey) {
            throw new Error(COMMON_CRYPTO_ERROR_NOPUBKEY)
          }

          return crypto.buildSignSuite({
            publicKey: key.pubKey,
            privateKey: key.pk || '',
            controller: options.controller,
            id: options.verificationMethod
          })
        },
        documentLoader
      })(credential)

      if (result.kind !== 'valid') {
        return [false, result]
      }

      return [true, result]
    },

    buildPresentation: async <
      C extends CommonCredentail = CommonCredentail,
      H extends CommonPresentationHolder = CommonPresentationHolder
    >(credentails: C[], options: CommonBuildPresentationOptions<H>) => {
      return buildVPV1Unsigned({
        id: `urn:uuid:${basicHelper.makeRandomUuid()}`,
        vcs: credentails,
        holder: options.holder,
        context: options.context,
        type: options.type
      }) as CommonUnsignedPresentation<C, H>
    }
  }
}