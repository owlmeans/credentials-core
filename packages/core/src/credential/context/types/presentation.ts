import { VPV1, VPV1Holder, VPV1Type, VPV1Unsigned } from "@affinidi/vc-common";
import { CommonCredentail } from "./credential";


export type CommonUnsignedPresentation<
  Credential extends CommonCredentail = CommonCredentail,
  Holder extends CommonPresentationHolder = CommonPresentationHolder
  > = VPV1Unsigned<Credential, VPV1Type, Holder>

export type CommonPresentation<
  Credential extends CommonCredentail = CommonCredentail,
  Holder extends CommonPresentationHolder = CommonPresentationHolder
  > = VPV1<Credential, VPV1Type, Holder>

export type CommonPresentationHolder = VPV1Holder & {
  did?: string
}

export type CommonPresentationType = VPV1Type