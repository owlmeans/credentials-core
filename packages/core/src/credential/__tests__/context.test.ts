require('dotenv').config()

import { buildCommonContext } from "../context"
import { CommonContext } from "../context/types"
import { Credential, UnsignedCredentail } from "../types"
import { nodeCryptoHelper } from "@owlmeans/regov-ssi-common"
import {
  buildDidHelper,
  buildDidRegistryWarpper,
  DIDDocument,
  DIDPURPOSE_ASSERTION,
  DIDPURPOSE_AUTHENTICATION,
  DIDPURPOSE_VERIFICATION
} from "@owlmeans/regov-ssi-did"
import { buildKeyChain } from "../../keys/model"

import util from 'util'
import { Presentation, UnsignedPresentation } from "../types"
util.inspect.defaultOptions.depth = 8


const test: {
  ctx?: CommonContext
  unsigned?: UnsignedCredentail
  credential?: Credential
  unsignedP?: UnsignedPresentation
  presentation?: Presentation
} = {}

beforeAll(async () => {
  test.ctx = await buildCommonContext({
    keys: await buildKeyChain({
      password: '11111111',
      crypto: nodeCryptoHelper
    }),
    crypto: nodeCryptoHelper,
    did: buildDidRegistryWarpper(buildDidHelper(nodeCryptoHelper))
  })
})

describe('Credential Context', () => {
  it('creates Credentail', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }

    const didWrapper = test.ctx.did
    const keys = test.ctx.keys

    const subject = {
      data: {
        '@type': 'TestCredentialSubjectDataType',
        worker: 'Valentin Michalych'
      }
    }

    const key = await keys.getCryptoKey()
    const didUnsigned = await didWrapper.helper().createDID(
      key,
      {
        data: JSON.stringify(subject),
        hash: true,
        purpose: [DIDPURPOSE_VERIFICATION, DIDPURPOSE_ASSERTION, DIDPURPOSE_AUTHENTICATION]
      }
    )

    const did = await didWrapper.helper().signDID(key, didUnsigned)
    expect(await didWrapper.helper().verifyDID(did)).toBe(true)

    didWrapper.addDID(did)

    const unsingnedCredentail = await test.ctx.buildCredential({
      id: did.id,
      type: ['VerifiableCredential', 'TestCredential'],
      holder: didWrapper.helper().extractProofController(did),
      context: {
        '@version': 1.1,
        meta: 'https://meta-id.meta-belarus.org/vc-schema#',
        data: {
          '@id': 'meta:data',
          '@type': '@id',
          '@context': {
            worker: { '@id': 'meta:worker', '@type': 'xsd:string' }
          }
        }
      },
      subject
    })

    test.unsigned = unsingnedCredentail
    expect(unsingnedCredentail).toMatchSnapshot({
      id: expect.any(String),
      holder: {
        id: expect.any(String)
      },
      issuanceDate: expect.any(String)
    })
  })

  it('signs Credential', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.unsigned) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const didWrapper = test.ctx.did
    const keys = test.ctx.keys

    const did = <DIDDocument>await didWrapper.lookUpDid(test.unsigned.id)
    const controller = didWrapper.helper().extractProofController(did)
    const key = await didWrapper.extractKey(controller)
    await keys.expandKey(key)

    const credentail = await test.ctx.signCredential(
      test.unsigned,
      controller,
      await keys.getCryptoKey()
    )

    test.credential = credentail

    expect(credentail).toMatchSnapshot({
      id: expect.any(String),
      holder: {
        id: expect.any(String)
      },
      issuer: expect.any(String),
      issuanceDate: expect.any(String),
      proof: {
        created: expect.any(String),
        jws: expect.any(String),
        verificationMethod: expect.any(String),
      }
    })
  })

  it('verifies Credential', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }
    const didWrapper = test.ctx.did
    const did = <DIDDocument>await didWrapper.lookUpDid(test.credential.id)
    const controller = didWrapper.helper().extractProofController(did)
    const key = await didWrapper.extractKey(controller)

    const [result, _] = await test.ctx.verifyCredential(test.credential, key)

    expect(result).toBe(true)
  })

  it('fails with the key of another issuer key', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const anotherContext = await buildCommonContext({
      keys: await buildKeyChain({
        password: '11111111',
        crypto: nodeCryptoHelper
      }),
      crypto: nodeCryptoHelper,
      did: buildDidRegistryWarpper(buildDidHelper(nodeCryptoHelper))
    })
    const cryptoKey = await anotherContext.keys.getCryptoKey()

    const [result] = await test.ctx.verifyCredential(
      test.credential,
      cryptoKey
    )

    expect(result).toBe(false)
  })

  it('fails with new issuer key', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    await test.ctx.keys.createKey('newKey')
    const cryptoKey = await test.ctx.keys.getCryptoKey('newKey')

    const [result] = await test.ctx.verifyCredential(
      test.credential,
      cryptoKey
    )

    expect(result).toBe(false)
  })

  it('doesn\'t allow to temper subject', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const newCredential = <Credential<{
      data: {
        '@type': string,
        worker: string
      }
    }>>{ ...test.credential }
    newCredential.credentialSubject = { ...newCredential.credentialSubject }
    newCredential.credentialSubject.data = { ...newCredential.credentialSubject.data }

    newCredential.credentialSubject.data.worker = `${newCredential.credentialSubject.data.worker}_`

    const [result] = await test.ctx.verifyCredential(
      newCredential,
      await test.ctx.keys.getCryptoKey()
    )

    expect(result).toBe(false)
  })

  it('creates unsigned verifiable presentation', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.credential) {
      throw 'Previous test didn\'t provide UnsingedCredential'
    }

    const didWrapper = test.ctx.did
    // In test we take did document of the credential itself
    const did = <DIDDocument>await didWrapper.lookUpDid(test.credential.holder.id)

    const vp = await test.ctx?.buildPresentation(
      [test.credential],
      {
        holder: did,
        type: 'TestPresentation'
      }
    )

    test.unsignedP = vp

    expect(vp).toMatchSnapshot(
      {
        id: expect.any(String),
        holder: {
          id: expect.any(String)
        },
        verifiableCredential: [
          {
            holder: {
              id: expect.any(String),
            },
            id: expect.any(String),
            issuanceDate: expect.any(String),
            issuer: expect.any(String),
            proof: {
              created: expect.any(String),
              jws: expect.any(String),
              verificationMethod: expect.any(String),
            }
          },
        ]
      }
    )
  })

  it('signs verifiable presentation', async () => {
    if (!test.ctx) {
      throw 'Setup didn\'t provide CommonContext'
    }
    if (!test.unsignedP) {
      throw 'Previous test didn\'t provide UnsingedPresentation'
    }
    const did = <DIDDocument>await test.ctx.did.lookUpDid(test.unsignedP.holder.id)
    if (!did) {
      throw 'No related did in registry'
    }

    const controller = test.ctx.did.helper().extractProofController(did)
    const key = await test.ctx.did.extractKey(`${controller}#${DIDPURPOSE_AUTHENTICATION}-1`)
    await test.ctx.keys.expandKey(key)

    // console.log(test.unsignedP)

    const vp = await test.ctx.signPresentation(test.unsignedP, did, key)

    // console.log(vp)
  })
})