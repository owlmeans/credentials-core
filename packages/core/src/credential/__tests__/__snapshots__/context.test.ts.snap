// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Credential Context creates Credentail 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "worker": Object {
            "@id": "meta:worker",
            "@type": "xsd:string",
          },
        },
        "@id": "meta:data",
        "@type": "@id",
      },
      "meta": "https://meta-id.meta-belarus.org/vc-schema#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "TestCredentialSubjectDataType",
      "worker": "Valentin Michalych",
    },
  },
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "issuanceDate": Any<String>,
  "type": Array [
    "VerifiableCredential",
    "TestCredential",
  ],
}
`;

exports[`Credential Context creates unsigned verifiable presentation 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "type": Array [
    "VerifiablePresentation",
    "TestPresentation",
  ],
  "verifiableCredential": Array [
    Object {
      "@context": Array [
        "https://www.w3.org/2018/credentials/v1",
        Object {
          "@version": 1.1,
          "data": Object {
            "@context": Object {
              "worker": Object {
                "@id": "meta:worker",
                "@type": "xsd:string",
              },
            },
            "@id": "meta:data",
            "@type": "@id",
          },
          "meta": "https://meta-id.meta-belarus.org/vc-schema#",
        },
      ],
      "credentialSubject": Object {
        "data": Object {
          "@type": "TestCredentialSubjectDataType",
          "worker": "Valentin Michalych",
        },
      },
      "holder": Object {
        "id": Any<String>,
      },
      "id": Any<String>,
      "issuanceDate": Any<String>,
      "issuer": Any<String>,
      "proof": Object {
        "created": Any<String>,
        "jws": Any<String>,
        "proofPurpose": "assertionMethod",
        "type": "EcdsaSecp256k1Signature2019",
        "verificationMethod": Any<String>,
      },
      "type": Array [
        "VerifiableCredential",
        "TestCredential",
      ],
    },
  ],
}
`;

exports[`Credential Context signs Credential 1`] = `
Object {
  "@context": Array [
    "https://www.w3.org/2018/credentials/v1",
    Object {
      "@version": 1.1,
      "data": Object {
        "@context": Object {
          "worker": Object {
            "@id": "meta:worker",
            "@type": "xsd:string",
          },
        },
        "@id": "meta:data",
        "@type": "@id",
      },
      "meta": "https://meta-id.meta-belarus.org/vc-schema#",
    },
  ],
  "credentialSubject": Object {
    "data": Object {
      "@type": "TestCredentialSubjectDataType",
      "worker": "Valentin Michalych",
    },
  },
  "holder": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "issuanceDate": Any<String>,
  "issuer": Any<String>,
  "proof": Object {
    "created": Any<String>,
    "jws": Any<String>,
    "proofPurpose": "assertionMethod",
    "type": "EcdsaSecp256k1Signature2019",
    "verificationMethod": Any<String>,
  },
  "type": Array [
    "VerifiableCredential",
    "TestCredential",
  ],
}
`;
